### PR TITLE
[SPARK-19758][SQL] Resolving timezone aware expressions with time zone when resolving inline table

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

When we resolve inline tables in analyzer, we will evaluate the expressions of inline tables.

When it evaluates a `TimeZoneAwareExpression` expression, an error will happen because the `TimeZoneAwareExpression` is not associated with timezone yet.

So we need to resolve these `TimeZoneAwareExpression`s with time zone when resolving inline tables.

## How was this patch tested?

Jenkins tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by viirya** [comment]
cc @ueshin

**Comment by rxin** [comment]
Put the test case in a sql file?

**Comment by viirya** [comment]
Ok. I will put it in inline-table.sql.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala [hunk 1]
```diff
@@ -95,10 +95,14 @@ object ResolveInlineTables extends Rule[LogicalPlan] {
...
-          if (e.dataType.sameType(targetType)) {
-            e.eval()
+          val castedExpr = if (e.dataType.sameType(targetType)) {
+            e
           } else {
-            Cast(e, targetType).eval()
+            Cast(e, targetType)
+          }
+          castedExpr match {
+            case te: TimeZoneAwareExpression => te.withTimeZone(conf.sessionLocalTimeZone).eval()
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala]: Shouldn't we traverse the entire tree expression to check for time zone aware expressions?

File: sql/core/src/test/resources/sql-tests/inputs/inline-table.sql [hunk 2]
```diff
@@ -29,6 +29,9 @@ select * from values ("one", array(0, 1)), ("two", array(2, 3)) as data(a, b);
 -- decimal and double coercion
 select * from values ("one", 2.0), ("two", 3.0D) as data(a, b);
 
+-- string to timestamp
+select * from values timestamp('1991-12-06 00:00:00.0') as data(a);
```
Comment (hvanhovell) [inline sql/core/src/test/resources/sql-tests/inputs/inline-table.sql]: Could you put this at the end of the file? That reduces the size of the diff.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala]: Or did `ResolveTimeZone` already process the input expression? In that case just add the timezone to the cast.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala]: We should. `ResolveTimeZone` is performed after all resolution rules finish.

**Review by hvanhovell — APPROVED** [review]
LGTM - pending jenkins

**Comment by viirya** [comment]
Thanks @hvanhovell @rxin

**Comment by hvanhovell** [comment]
LGTM - merging to master. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -146,7 +146,7 @@ class Analyzer(
       GlobalAggregates ::
       ResolveAggregateFunctions ::
       TimeWindowing ::
-      ResolveInlineTables ::
+      ResolveInlineTables(conf) ::
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: is it possible to fix this bug by re-order analyzer rules?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I tried before. But the resolution rules will add new timezone-aware expressions, so it still needs the rule to resolve timezone-aware expressions after resolution rules.

**Pull request closed without merging**